### PR TITLE
Track 3: K-Maxwell momentum 3160-step on a Muon tuned baseline result (n=8)

### PR DESCRIPTION
**Pull request by jacknzheng**
# Track 3: annealed K-Maxwell momentum on tuned Muon — 3160 steps (n=8)

Collaborator: [Jeffrey Cheng](https://github.com/jeffreycider) ([@jeffreycider](https://github.com/jeffreycider)).

The tuned Muon + aux AdamW baseline (result #36) with one optimizer change: replace bi-Maxwell momentum with **K=8 log-spaced EMA buffers**, then linearly anneal their mixture mean age from **58 → 26 steps**.

Read the [full report in the record README](https://github.com/jacknzheng/kmaxwell-sota/blob/fb2536fbdf26dac960709de56edc8ff85c966d6a/records/track_3_optimization/results/20260826_kmaxwell_3160/README.md).

## Result

- 8 seeds (`0–7`, H100)
- First statistically passing step: **3160**
- Mean val loss: `3.27794`
- `(3.28 - mean) * sqrt(8) = 0.00584 ≥ 0.004`
- Step 3150 fails (`0.00360`)
- **−90 steps** versus the #36 Muon baseline (3250)
- Pairwise significance versus #36: `0.0122` at 3200 and `0.0113` at 3250

Everything else—architecture, data, batch size, learning-rate schedule, weight decay, and aux AdamW—is unchanged.

## Change

```python
# per Muon 2-D parameter, step >= 1000
for k, beta in enumerate(kmaxwell_decay_rates):
    m[k].lerp_(g, 1 - beta)

frac = (step - 1000) / 2250
w = (1 - frac) * w_age58 + frac * w_age26
m_eff = sum(w[k] * m[k] for k in range(8))
update = g.lerp_(m_eff, mu)
```

All buffers lazy-initialize from the existing momentum at step 1000, preserving the switch-step update.

## Curves

full descent

target zone

## Other contributors

Thank you to [Jerry Hong](https://github.com/jerryhong21) ([@jerryhong21](https://github.com/jerryhong21)) for assisting us with this project.

## Setup & acknowledgements

Baseline: result #36 (tuned Muon + aux AdamW) by [@konstmish](https://github.com/konstmish) (PR [#323](https://github.com/KellerJordan/modded-nanogpt/pull/323)); `train_gpt_simple.py` by [@kellerjordan0](https://x.com/kellerjordan0). Related momentum-mixture prior work: AggMo (Lucas et al., 2018), QHM (Ma & Yarats, 2019), AdEMAMix (Pagliardini et al., 2024), and the two-timescale bi-Maxwell kernel on this same baseline ([PR #340](https://github.com/KellerJordan/modded-nanogpt/pull/340), 3210). This entry differs in using K>2 log-spaced ticks with a **time-varying** mean age.

### CODE REVIEW

**Comment by Hongyu-yu** [comment]
Happy to see K-Maxwell works better! Really nice further work! Our Bi-Maxwell arXiv preprint is online now at:
https://arxiv.org/abs/2608.22994
https://www.alphaxiv.org/abs/2608.22994?utm_source=extension&utm_medium=abstract-page
Hope it interests you.
